### PR TITLE
CMP #1: Add `process` method to `MeasurementProcess` class

### PR DESCRIPTION
**Pull request by AlbertMitjans**
`SampleMeasurement` and `StateMeasurement` classes have been added. They contain an abstract
  method to process samples/quantum state.

All measurement processes will inherit from one/both classes and define its own `process`/`process_state` method, which will process the input samples/quantum state.

### CODE REVIEW

File: pennylane/measurements/measurements.py
Comment (eddddddy) [inline pennylane/measurements/measurements.py:723]: Is there a reason this has a different name? Seems like it would make more sense for these names to be the same.

File: pennylane/measurements/measurements.py
Comment (eddddddy) [inline pennylane/measurements/measurements.py:705]: This is up for debate, but IMO I think it would make more sense for this method to take a `device` argument instead of what's given here. One advantage for this is that the signature for this function is the same across all measurement processes. My other comment is also relevant here.

File: pennylane/measurements/measurements.py
Comment (AlbertMitjans) [reply pennylane/measurements/measurements.py:723]: Some measurements will inherit from both `SampleMeasurement` and `StateMeasurement` classes. Given that we cannot perform method overloading in python, we need to use different names.

Comment (AlbertMitjans) [reply pennylane/measurements/measurements.py:723]: When supporting these methods in the device, we'll have something like the following:
```python
class Device:
    def statistics(self, measurements):
        results = []
        for m in measurements:
            if self.shots is None:
                results.append(m.process_state(self.state, self.wires)
            else:
                results.append(m.process(self._samples, shot_range, bin_size)
```

File: pennylane/measurements/measurements.py
Comment (AlbertMitjans) [reply pennylane/measurements/measurements.py:705]: But as I mentioned last week, the `classical_shadow` measurement is a special case. Honestly I believe it should be considered as a tape transforms instead of a measurement: the tape transform would generate `n_shots` tapes with `shots=1`, execute them on the device, and then generate the outcomes and recipes in the classical post processing.

Comment (AlbertMitjans) [reply pennylane/measurements/measurements.py:705]: > To break the coupling between measurements and devices we need to define a set of arguments/variables that are sufficient for computing any measurement process.

This is a similar approach to the one followed in the new device API refactor: they want to introduce a class `ExecutionConfig` that will contain all the needed information for the execution of a tape.

File: pennylane/measurements/measurements.py
Comment (eddddddy) [inline pennylane/measurements/measurements.py:733]: Unused arguments.

Comment (Jaybsoni) [reply pennylane/measurements/measurements.py:733]: +1 to update doc string.

**Review by Jaybsoni — APPROVED** [review]
Looks good to me, just need to update one doc-string

File: pennylane/measurements/measurements.py
Comment (AlbertMitjans) [reply pennylane/measurements/measurements.py:705]: I agree with you that right now it would make more sense for the device to handle all the shot information. However in the new device API we are planning on moving shots into an "`ExecutionConfig`" class that will be passed around. That is why I decided to also pass the shots information into the measurement class, to comply with the new refactor.

File: pennylane/measurements/measurements.py
Comment (AlbertMitjans) [reply pennylane/measurements/measurements.py:716]: Good call! Done.

File: pennylane/measurements/measurements.py
Comment (AlbertMitjans) [reply pennylane/measurements/measurements.py]: Updated

**Comment by AlbertMitjans** [comment]
Hi @eddddddy! I will wait for your approval to merge this PR.

File: pennylane/measurements/measurements.py
Comment (eddddddy) [reply pennylane/measurements/measurements.py:716]: It still prevents calls like `super().process()` from the subclass. Anyways, this comment is minor.

**Review by eddddddy — APPROVED** [review]

File: pennylane/measurements/measurements.py
Comment (AlbertMitjans) [reply pennylane/measurements/measurements.py:716]: I see what you mean. If you don't mind, I will add the `NotImplementedError` in PR#2, because then I can test that `super().process_samples` raises an error.

**Pull request merged**
